pwndbg/commands/heap.py

```diff
@@ -549,49 +549,96 @@ def tcachebins(addr=None, verbose=False):
-    start = int(addr) - max_fast + psize
-    if start < 0:
+
+    if size is None:
+        size = max_fast
+    elif size > addr:
+        print(message.warn("Size of 0x%x is greater than the target address 0x%x", (size, addr)))
+        size = addr
+    elif size > max_fast:
+        print(
+            message.warn(
+                "0x%x is greater than the global_max_fast value of 0x%x" % (size, max_fast)
+            )
+        )
+    elif size < min_fast:
         print(
             message.warn(
-                "addr - global_max_fast is negative, if the max_fast is not corrupted, you gave wrong address"
+                "0x%x is smaller than the minimum fastbin chunk size of 0x%x" % (size, min_fast)
             )
         )
-        start = 0  # TODO, maybe some better way to handle case when global_max_fast is overwritten with something large
-    mem = pwndbg.gdblib.memory.read(start, max_fast - psize, partial=True)
+        size = min_fast
 
-    fmt = {"little": "<", "big": ">"}[pwndbg.gdblib.arch.endian] + {4: "I", 8: "Q"}[psize]
+    start = int(addr) - size + psize
 
-    if size is None:
-        sizes = range(min_fast, max_fast + 1, align)
-    else:
-        sizes = [int(size)]
+    if align:
+        # If a chunk is aligned to MALLOC_ALIGNMENT, the size field should be at
+        # offset `psize`. First we align up to a multiple of `psize`
+        new_start = pwndbg.lib.memory.align_up(start, psize)
+
+        # Then we make sure we're at a multiple of `psize` but not `psize*n` for
+        # n > 1, by making sure the bottom nibble gets set to `psize`
+        new_start |= psize
+
+        # If we're starting at a higher address, we still only want to read
+        # enough bytes to reach our target address
+        size -= new_start - start
+
+        start = new_start
+
+    print(
+        message.notice(
+            "Searching for fastbin sizes up to 0x%x starting at 0x%x resulting in an overlap of 0x%x"
+            % (size, start, addr)
+        )
+    )
+
+    # Only consider `size - psize` bytes, since we're starting from after `prev_size`
+    mem = pwndbg.gdblib.memory.read(start, size - psize, partial=True)
+
+    fmt = {"little": "<", "big": ">"}[pwndbg.gdblib.arch.endian] + {4: "I", 8: "Q"}[psize]
 
     print(C.banner("FAKE CHUNKS"))
-    for size in sizes:
-        fastbin = allocator.fastbin_index(size)
-        for offset in range((max_fastbin - fastbin) * align, max_fast - align + 1):
-            candidate = mem[offset : offset + psize]
-            if len(candidate) == psize:
-                value = struct.unpack(fmt, candidate)[0]
-                if allocator.fastbin_index(value) == fastbin:
-                    malloc_chunk(start + offset - psize, fake=True)
+    step = malloc_alignment if align else 1
+    for i in range(0, len(mem), step):
+        candidate = mem[i : i + psize]
+        if len(candidate) == psize:
+            value = struct.unpack(fmt, candidate)[0]
```

Comment:
Random thought: instead of `memory.read(...)` we could do `memory.pvoid(...)` probably

Reply:
It would work, but it should be more efficient to read all of the memory at once instead of one pointer at a time.

Reply:
I guess it depends on the memory size and target, e.g. if you debug a remote target via additional debugging hardware (jtag?) then memory accesses may be slow and it may be better to read 100B instead of 1kB etc (?). Though I may be wrong and maybe reading a big blob will be faster.

Reply:
Anyway, I'm fine with reading all at once. We could optionally add a comment that it may be worth changing this for remote debugging but we do not know if its worth it (?).